src/java/org/apache/cassandra/cql3/statements/SelectStatement.java

```diff
@@ -1485,35 +1537,30 @@ private ColumnComparator<List<ByteBuffer>> getOrderingComparator(Selection selec
                     : new CompositeComparator(sorters, idToSort);
         }
 
-        private boolean isReversed(TableMetadata table, Map<ColumnMetadata, Ordering> orderingColumns, StatementRestrictions restrictions) throws InvalidRequestException
+        private boolean isReversed(TableMetadata table, Map<ColumnMetadata, Ordering> orderingColumns) throws InvalidRequestException
         {
-            // Nonclustered ordering handles descending logic in a different way
+            // Nonclustered ordering handles descending logic through ScoreOrderedResultRetriever and TKP
             if (orderingColumns.values().stream().anyMatch(o -> o.expression.hasNonClusteredOrdering()))
                 return false;
 
-            Boolean[] reversedMap = new Boolean[table.clusteringColumns().size()];
-            int i = 0;
+            Boolean[] clusteredMap = new Boolean[table.clusteringColumns().size()];
             for (var entry : orderingColumns.entrySet())
             {
                 ColumnMetadata def = entry.getKey();
                 Ordering ordering = entry.getValue();
-                boolean reversed = ordering.direction == Ordering.Direction.DESC;
-
-                // VSTODO move this to verifyOrderingIsAllowed?
-                checkTrue(def.isClusteringColumn(),
-                          "Order by is currently only supported on the clustered columns of the PRIMARY KEY, got %s", def.name);
-                while (i != def.position())
-                {
-                    checkTrue(restrictions.isColumnRestrictedByEq(table.clusteringColumns().get(i++)),
-                              "Order by currently only supports the ordering of columns following their declared order in the PRIMARY KEY");
-                }
-                i++;
-                reversedMap[def.position()] = (reversed != def.isReversedType());
+                // We defined ANN OF to be ASC ordering, as in, "order by near-ness".  But since score goes from
+                // 0 (worst) to 1 (closest), we need to reverse the ordering for the comparator when we're sorting
+                // by synthetic +score column.
+                boolean cqlReversed = ordering.direction == Ordering.Direction.DESC;
+                if (def.position() == ColumnMetadata.NO_POSITION)
+                    return ordering.expression.isScored() || cqlReversed;
```

Comment:
I don't think this hurts anything as is, but is the "isScored()" ever going to return true?

Reply:
Ann and BM25 Ordering subclasses override isScored() to true.  Maybe I'm not understanding your point correctly.